libcontextprovider/context-provide/context-provide.1
.TH context-provide 1 Nov-18-2009

.SH NAME
context-provide
- A tool to provide ContextKit properties to debug and inspect applications

.SH SYNOPSIS
.B context-provide [\fI--session\fR | \fI--system\fR] \fR[\fIBUSNAME\fR] [\fITYPE\fR \fINAME\fR \fIINITVALUE\fR]...

.SH DESCRIPTION
\fBcontext-provide\fR is an interactive tool to set (provide)
properties of the ContextKit.  It can be used to simulate an actual
provider to debug applications using the ContextKit.  When launched,
the tool presents a simple console where commands can be issued to
provide new keys and set their values.

Properties can also be provided on the command line with \fITYPE\fR
\fINAME\fR \fIINITVALUE\fR triplets.

When ran, the tool automatically claims the specified \fIBUSNAME\fR and if
there were properties specified on the command line, it will provide them
right away.  To add new properties, use the \fB`add'\fR command.

If no \fIBUSNAME\fR is given the tool acts as a Commander, taking control of
all subscribers in the system.  This mode by default proxies the real values
of all properties.  To control a property you have to \fB`add'\fR the
property, which stops forwarding the corresponding property.  You can undo
this via the \fB`del'\fR command.

Values are input using JSon notation (http://json.org/) relaxed to allow other
top-level elements than objects (ie. you can write `"some text"' or `23.1'
directly).  As an executive summary, the allowed data types are \fBstrings\fR
("soylent green"), \fBnumbers\fR (2, 6.9), heterogeneous \fBlists\fR ([1, 1.5,
3, "foo"]) and \fBdictionaries\fR with string keys ({"foo": 1, "bar": 2}).

.SH OPTIONS
.TP
\fB--system\fR
Use the system bus.
.TP
\fB--session\fR
Use the session bus (the default).

.SH COMMANDS
.TP
\fBadd\fR \fITYPE\FR \fIKEY\fR [\fIINITVALUE\fR]

Adds a key named \fIKEY\fR with the given \fITYPE\fR to the list of provided
properties.  The initial value will be \fIINITVALUE\fR if specified, otherwise
unset (null).  This has to be called before an attempt is made to set a key
value.  \fITYPE\fR can be any type described in the \fBtype registry\fR
(located in /usr/share/contextkit/core.types, overridden by the
CONTEXT_CORE_TYPES environment variable).  Some examples are: value, bool,
number, integer, string, list, map.  For backward compatibility, INT, BOOL,
and DOUBLE may also be used as type.

Example: "add integer Battery.Status 20".

In \fICommander mode\fR it also stops proxying the real property.
.TP
\fBsettype\fR \fIKEY\fR \fITYPE FRAGMENT\fR

Sets the type of \fIKEY\fR to the NanoXML fragment.  For convenience, "type"
is interpreted as "<type/>".  See
/usr/share/doc/contextkit-doc/html/context-types.html for more information.
.TP
\fBdel\fR \fIKEY\fR

Makes context-provide forget \fIKEY\fR.  In \fICommander mode\fR, restores
proxying of the real property.
.TP
\fIKEY\fB=\fIVALUE\fR

Sets the given \fIKEY\fR value to the new \fIVALUE\fR, which has to be in JSon
format (see above).  For backward compatibility, if the type of the property
is \fBstring\fR, the value is accepted without quotes.

Example: "Battery.Status=99".

.TP
\fBunset\fR \fIKEY\fR

Sets the given \fIKEY\fR to non specified (null).
.TP
\fBinfo\fR \fIKEY\fR

Prints the type, name and current value of \fIKEY\fR.  In \fICommander mode\fR
the real value (from the original provider) is printed as well.
.TP
\fBlist\fR

Same as calling `info' for all known keys (both proxied and added ones).
.TP
\fBsleep\fR \fIINTERVAL\fR

Sleeps and blocks the main loop for the given amount of seconds.  Used mainly
for internal testing purposes.  Example: "sleep 10".
.TP
\fBdump\fR [\fIFILENAME\fR]

Dumps the currently added properties to an XML .context file defined by the
\fIFILENAME\fR (the default is 'context-provide.context' in the current
directory).
.TP
\fBrestart\fR

Tries to republish the \fIBUSNAME\fR of the provider on D-Bus, exit if it
fails.
.TP
\fBexit\fR

Exits the program.

.SH EXAMPLE SESSIONS
Starts up a provider, initially having a single integer property named
an.example with the default value of 42.

  ./context-provide dbus.service.name integer an.example 42

Starts up the provider on the system bus without any properties, then
add some through the command line interface.

  ./context-provide --system org.bottomdrawer.kit2000
    add integer example.int 33
    add string example.string
    example.string = value1
    example.string = value2
    unset example.int

.SH SEE ALSO
.BR context-listen (1)
.BR context-ls (1)
.BR /usr/share/doc/contextkit-doc/html/*.html


NEWS
See debian/changelog for news.


libcontextprovider/context-provide/context-provide.cpp
/*
 * Copyright (C) 2009 Nokia Corporation.
 *
 * Contact: Marius Vollmer <marius.vollmer@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 */

#include "commandwatcher.h"

#include <QCoreApplication>
#include <QString>
#include <QStringList>
#include <QMap>
#include <QDebug>
#include <stdlib.h>
#include <service.h>

using namespace ContextProvider;

int main(int argc, char **argv)
{
    QCoreApplication app(argc, argv);

    QStringList args = app.arguments();
    QString busName;
    QDBusConnection::BusType busType = QDBusConnection::SessionBus;
    QTextStream out(stdout);

    if (args.contains("--help") || args.contains("-h")) {
        // Help? Show it and be gone.
        out << "Usage: " << argv[0] << " [--session | --system] [BUSNAME]\n";
        out << "BUSNAME is " << CommandWatcher::commanderBusName << " by default, and bus is session.\n";
        return 0;
    }

    // Silently dropping --v2, which might be still used by some tests.
    if (args.contains("--v2"))
        args.removeAll("--v2");

    // session/system
    if (args.contains("--session")) {
        busType = QDBusConnection::SessionBus;
        args.removeAll("--session");
    }

    if (args.contains("--system")) {
        busType = QDBusConnection::SystemBus;
        args.removeAll("--system");
    }

    bool readInput = true;
    if (args.contains("--no-input")) {
        args.removeAll("--no-input");
        readInput = false;
    }

    if (args.count() < 2) {
        // No arguments at all? Use commander by default.
        args.push_back(CommandWatcher::commanderBusName);
    }

    // Parameter? Extract the session bus and type from it.
    busName = args.at(1);

    Service service(busType, busName, false);
    service.setAsDefault();

    // 0 -> prog name
    // 1 -> busname
    // 2 -> type of prop1
    // 3 -> name of prop1
    // 4 -> value of prop1
    // 5 -> type of prop2
    // 6 -> name of prop2
    // 7 -> value of prop2
    // ...
    if (args.count() % 3 != 2) {
        qDebug() << "Wrong number of properties\n";
        return 1;
    }

    qDebug() << "Service:" << busName.toLocal8Bit().data() << "on" <<
        ((busType == QDBusConnection::SessionBus) ? "session" : "system");

    CommandWatcher commandWatcher(busName, busType, STDIN_FILENO, readInput,
                                  QCoreApplication::instance());

    for (int i=2; i < args.count(); i+=3)
        commandWatcher.addCommand(args.mid(i, 3));

    commandWatcher.restartCommand();

    return app.exec();
}
